docs/TRACEBACKS.md

```diff
@@ -0,0 +1,200 @@
+    first, second = [1]
+ValueError: not enough values to unpack (expected 2, got 1)
+
+```
+
+It says that the most recent call is last, so the place to start is with the exception at the bottom of the trace.
+From there it can be traced back to the start.
+If we place the offending line in a function and then call the function we see a longer trace:
+
+```python
+>>> def my_func():
+...     first, second = [1]
+...
+>>> my_func()
+Traceback (most recent call last):
+  File <stdin>, line 5, in <module>
+    my_func()
+  File <stdin>, line 2, in my_func
+    first, second = [1]
+ValueError: not enough values to unpack (expected 2, got 1)
+
+```
+
+Working backwards from the bottom, we see that the call where the exception happened is on line 2 in `my_func`.
+We got there by calling `my_func` on line 5.
+
+## Using the `print` function
+
+Sometimes an error is not being raised, but a value is not what is expected.
+This can be especially perplexing if the value is the result of a chain of calculations.
+In such a situation it can be helpful to look at the value at each step to see which step is the one that isn't behaving as expected.
+The [print][print] function can be used for printing the value to the console.
+The following is an example of a function that doesn't return the value expected:
+
+```python
+# the intent is to only pass an int to this function and get an int back
+def halve_and_quadruple(num):
+    return (num / 2) * 4
+```
+
+When the function is passed `5`, the expected value is `8`, but it it returns `10.0`.
+To troubleshoot, the calculating is broken up so that the value can be inspected at every step.
+
+
+```python
+# the intent is to only pass an int to this function and get an int back
+def halve_and_quadruple(num):
+    # verify the number in is what is expected
+    # prints 5
+    print(num)
+    # we want the int divided by an int to be an int
+    # but this prints 2.5! We've found our mistake.
+    print(num / 2)
+    # this makes sense, since 2.5 x 4 = 10.0
+    print((num / 2) * 4)
+    return (num / 2) * 4
+
+```
+
+What the `print` calls revealed is that we used `/` when we should have used `//`, the [floor divison operator][floor divison operator].
+
+## Logging
+
+[Logging][logging] can be used similarly to `print`, but it is more powerful.
+What is logged can be configured by the logging severity (e.g., 'DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL'.)
+A call to the `logging.error` function can pass `True` to the `exc_info` parameter, which will additionally log the stack trace.
+By configuring multiple handlers, logging can print to more than one place with the same logging function.
+
+Following is a an example of logging printed to the console:
+
+```python
+
+>>> import logging
+...
+... # configures minimum logging level as INFO
+... logging.basicConfig(level=logging.INFO)
+
+... def halve_and_quadruple(num):
+...     # prints INFO:root: num == 5
+...     logging.info(f" num == {num}")
```

Comment:
Some linters will complain that you should use format strings to allow for lazy formatting.